File: app/src/main/java/igrek/songbook/model/lyrics/LyricsParser.java
package igrek.songbook.model.lyrics;

import android.graphics.Paint;
import android.graphics.Typeface;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class LyricsParser {
	
	private boolean bracket;
	private Paint paint;
	private final Set<String> wordSplitters = new HashSet<>(Arrays.asList(" ", ".", ",", "-", ":", ";", "/"));
	
	public LyricsParser() {
	}
	
	public synchronized LyricsModel parseFileContent(String content, float screenW, float fontsize, Paint paint) {
		this.paint = paint;
		setNormalFont();
		paint.setTextSize(fontsize);
		
		content = content.replace("\r", "");
		content = content.replace("\t", " ");
		LyricsModel model = new LyricsModel();
		setBracket(false);
		String[] lines1 = content.split("\n");
		for (String line1 : lines1) {
			model.addLines(parseLine(line1, screenW, fontsize));
		}
		
		// store line numbers
		int y = 0;
		for (LyricsLine line : model.getLines()) {
			line.setY(y);
			y++;
		}
		
		return model;
	}
	
	private List<LyricsLine> parseLine(String line, float screenW, float fontsize) {
		
		List<LyricsChar> chars = str2chars(line.trim());
		
		List<List<LyricsChar>> lines2 = wrapLine(chars, screenW);
		
		List<LyricsLine> lines = new ArrayList<>();
		for (List<LyricsChar> subline : lines2) {
			lines.add(chars2line(subline, fontsize));
		}
		
		return lines;
	}
	
	private List<LyricsChar> str2chars(String line) {
		List<LyricsChar> chars = new ArrayList<>();
		for (int i = 0; i < line.length(); i++) {
			String c = Character.toString(line.charAt(i));
			
			float charWidth;
			LyricsTextType type;
			
			if (c.equals("[")) {
				setBracket(true);
				charWidth = 0;
				type = LyricsTextType.BRACKET;
			} else if (c.equals("]")) {
				setBracket(false);
				charWidth = 0;
				type = LyricsTextType.BRACKET;
			} else {
				float[] fw = new float[1];
				paint.getTextWidths(c, fw);
				charWidth = fw[0];
				if (bracket) {
					type = LyricsTextType.CHORDS;
				} else {
					type = LyricsTextType.REGULAR_TEXT;
				}
			}
			
			chars.add(new LyricsChar(c, charWidth, type));
		}
		return chars;
	}
	
	
	private float textWidth(List<LyricsChar> chars) {
		float sum = 0;
		for (LyricsChar achar : chars) {
			sum += achar.width;
		}
		return sum;
	}
	
	private int maxScreenStringLength(List<LyricsChar> chars, float screenW) {
		int l = chars.size();
		while (textWidth(chars.subList(0, l)) > screenW && l > 1) {
			l--;
		}
		// do not wrap in the middle of the word, try to step back until word splitter found
		int lastWordSplitter = findLastWordSplitter(chars, l);
		if (lastWordSplitter == -1) {
			// it's one long word only - no way to split
			return l;
		} else if (lastWordSplitter == l - 1) {
			// last char is already a word splitter
			return l;
		} else {
			// split after last word splitter
			return lastWordSplitter + 1;
		}
	}
	
	private int findLastWordSplitter(List<LyricsChar> chars, int toIndex) {
		while (--toIndex > 1) {
			// is word splitter
			if (wordSplitters.contains(chars.get(toIndex).c))
				return toIndex;
		}
		return -1;
	}
	
	private List<List<LyricsChar>> wrapLine(List<LyricsChar> chars, float screenW) {
		List<List<LyricsChar>> lines = new ArrayList<>();
		if (textWidth(chars) <= screenW) {
			lines.add(chars);
		} else {
			int maxLength = maxScreenStringLength(chars, screenW);
			List<LyricsChar> before = chars.subList(0, maxLength);
			// copying due to subsequent modifications
			ArrayList<LyricsChar> newBefore = new ArrayList<>(before);
			// add special line wrapper
			newBefore.add(new LyricsChar("\u21B5", 0, LyricsTextType.LINEWRAPPER));
			List<LyricsChar> after = chars.subList(maxLength, chars.size());
			lines.add(newBefore);
			lines.addAll(wrapLine(after, screenW));
		}
		return lines;
	}
	
	private synchronized LyricsLine chars2line(List<LyricsChar> chars, float fontsize) {
		// aggregate groups of the same type
		LyricsLine line = new LyricsLine();
		
		LyricsTextType lastType = null;
		StringBuilder buffer = new StringBuilder();
		float startX = 0;
		float x = 0;
		for (int i = 0; i < chars.size(); i++) {
			LyricsChar lyricsChar = chars.get(i);
			if (lastType == null)
				lastType = lyricsChar.type;
			
			if (lyricsChar.type != lastType) {
				// complete the previous fragment
				if (buffer.length() > 0) {
					
					if (lastType.isDisplayable()) {
						LyricsFragment fragment = new LyricsFragment(startX / fontsize, buffer.toString(), lastType);
						line.addFragment(fragment);
					}
					
					startX = x;
					buffer = new StringBuilder();
				}
				
				lastType = lyricsChar.type;
			}
			
			if (lyricsChar.type.isDisplayable()) {
				buffer.append(lyricsChar.c);
				x += lyricsChar.width;
			}
		}
		
		// complete the last fragment
		if (buffer.length() > 0) {
			if (lastType.isDisplayable()) {
				LyricsFragment fragment = new LyricsFragment(startX / fontsize, buffer.toString(), lastType);
				line.addFragment(fragment);
			}
		}
		
		return line;
	}
	
	
	private void setBracket(boolean bracket) {
		this.bracket = bracket;
		// change typeface due to text width calculation
		if (bracket) {
			setBoldFont();
		} else {
			setNormalFont();
		}
	}
	
	private void setNormalFont() {
		paint.setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.NORMAL));
	}
	
	private void setBoldFont() {
		paint.setTypeface(Typeface.create(Typeface.DEFAULT, Typeface.BOLD));
	}
}

File: app/src/main/java/igrek/songbook/model/lyrics/LyricsParser.kt
package igrek.songbook.model.lyrics

import android.graphics.Paint
import android.graphics.Typeface
import java.util.*

class LyricsParser(fontFamily: Typeface) {

    private var bracket: Boolean = false
    private var paint: Paint? = null
    private val wordSplitters = hashSetOf(" ", ".", ",", "-", ":", ";", "/")

    private val normalTypeface: Typeface = Typeface.create(fontFamily, Typeface.NORMAL)
    private val boldTypeface: Typeface = Typeface.create(fontFamily, Typeface.BOLD)

    @Synchronized
    fun parseFileContent(content: String, screenW: Float, fontsize: Float, paint: Paint): LyricsModel {
        this.paint = paint
        setNormalFont()
        paint.textSize = fontsize

        val model = LyricsModel()
        setBracket(false)
        val lines1 = content.split("\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (line1 in lines1) {
            model.addLines(parseLine(line1, screenW, fontsize))
        }

        // store line numbers
        for ((y, line) in model.lines.withIndex()) {
            line.y = y
        }

        return model
    }

    private fun parseLine(line: String, screenW: Float, fontsize: Float): List<LyricsLine> {

        val chars = str2chars(line)

        val lines2 = wrapLine(chars, screenW)

        val lines = ArrayList<LyricsLine>()
        for (subline in lines2) {
            lines.add(chars2line(subline, fontsize))
        }

        return lines
    }

    private fun str2chars(line: String): List<LyricsChar> {
        val chars = ArrayList<LyricsChar>()
        for (i in 0 until line.length) {
            val c = Character.toString(line[i])

            val charWidth: Float
            val type: LyricsTextType

            when (c) {
                "[" -> {
                    setBracket(true)
                    charWidth = 0f
                    type = LyricsTextType.BRACKET
                }
                "]" -> {
                    setBracket(false)
                    charWidth = 0f
                    type = LyricsTextType.BRACKET
                }
                else -> {
                    val fw = FloatArray(1)
                    paint!!.getTextWidths(c, fw)
                    charWidth = fw[0]
                    type = if (bracket) {
                        LyricsTextType.CHORDS
                    } else {
                        LyricsTextType.REGULAR_TEXT
                    }
                }
            }

            chars.add(LyricsChar(c, charWidth, type))
        }
        return chars
    }


    private fun textWidth(chars: List<LyricsChar>): Float {
        var sum = 0f
        for (achar in chars) {
            sum += achar.width
        }
        return sum
    }

    private fun maxScreenStringLength(chars: List<LyricsChar>, screenW: Float): Int {
        var l = chars.size
        while (textWidth(chars.subList(0, l)) > screenW && l > 1) {
            l--
        }
        // do not wrap in the middle of the word, try to step back until word splitter found
        val lastWordSplitter = findLastWordSplitter(chars, l)
        return when (lastWordSplitter) {
            -1 -> l// it's one long word only - no way to split
            l - 1 -> l// last char is already a word splitter
            else -> lastWordSplitter + 1 // split after last word splitter
        }
    }

    private fun findLastWordSplitter(chars: List<LyricsChar>, _toIndex: Int): Int {
        var toIndex = _toIndex
        while (--toIndex > 1) {
            // is word splitter
            if (wordSplitters.contains(chars[toIndex].c))
                return toIndex
        }
        return -1
    }

    private fun wrapLine(chars: List<LyricsChar>, screenW: Float): List<List<LyricsChar>> {
        val lines = ArrayList<List<LyricsChar>>()
        if (textWidth(chars) <= screenW) {
            lines.add(chars)
        } else {
            val maxLength = maxScreenStringLength(chars, screenW)
            val before = chars.subList(0, maxLength)
            // copying due to subsequent modifications
            val newBefore = ArrayList(before)
            // add special line wrapper
            newBefore.add(LyricsChar("\u21B5", 0f, LyricsTextType.LINEWRAPPER))
            val after = chars.subList(maxLength, chars.size)
            lines.add(newBefore)
            lines.addAll(wrapLine(after, screenW))
        }
        return lines
    }

    @Synchronized
    private fun chars2line(chars: List<LyricsChar>, fontsize: Float): LyricsLine {
        // aggregate groups of the same type
        val line = LyricsLine()

        var lastType: LyricsTextType? = null
        var buffer = StringBuilder()
        var startX = 0f
        var x = 0f
        for (i in chars.indices) {
            val lyricsChar = chars[i]
            if (lastType == null)
                lastType = lyricsChar.type

            if (lyricsChar.type != lastType) {
                // complete the previous fragment
                if (buffer.isNotEmpty()) {

                    if (lastType!!.isDisplayable) {
                        val fragment = LyricsFragment(startX / fontsize, buffer.toString(), lastType)
                        line.addFragment(fragment)
                    }

                    startX = x
                    buffer = StringBuilder()
                }

                lastType = lyricsChar.type
            }

            if (lyricsChar.type.isDisplayable) {
                buffer.append(lyricsChar.c)
                x += lyricsChar.width
            }
        }

        // complete the last fragment
        if (buffer.isNotEmpty()) {
            if (lastType != null && lastType.isDisplayable) {
                val fragment = LyricsFragment(startX / fontsize, buffer.toString(), lastType)
                line.addFragment(fragment)
            }
        }

        return line
    }


    private fun setBracket(bracket: Boolean) {
        this.bracket = bracket
        // change typeface due to text width calculation
        if (bracket) {
            setBoldFont()
        } else {
            setNormalFont()
        }
    }

    private fun setNormalFont() {
        paint!!.typeface = normalTypeface
    }

    private fun setBoldFont() {
        paint!!.typeface = boldTypeface
    }
}
